pkg/storegateway/bucket_index_postings.go

```diff
@@ -29,54 +29,131 @@ type postingGroup struct {
 	removeKeys []labels.Label
 }
 
-func newPostingGroup(addAll bool, addKeys, removeKeys []labels.Label) *postingGroup {
-	return &postingGroup{
+func newPostingGroup(addAll bool, addKeys, removeKeys []labels.Label) postingGroup {
+	return postingGroup{
 		addAll:     addAll,
 		addKeys:    addKeys,
 		removeKeys: removeKeys,
 	}
 }
 
-// NOTE: Derived from tsdb.postingsForMatcher. index.Merge is equivalent to map duplication.
-func toPostingGroup(lvr indexheader.Reader, m *labels.Matcher) (*postingGroup, error) {
-	if setMatches := m.SetMatches(); len(setMatches) > 0 && (m.Type == labels.MatchRegexp || m.Type == labels.MatchNotRegexp) {
-		keys := make([]labels.Label, 0, len(setMatches))
-		for _, val := range setMatches {
-			if _, err := lvr.PostingsOffset(m.Name, val); errors.Is(err, indexheader.NotFoundRangeErr) {
+func (g postingGroup) filterNonExistentKeys(preader postingsOffsetReader) (_ postingGroup, retErr error) {
+	filterKeys := func(keys []labels.Label) []labels.Label {
+		existingKeys := 0
+		for i, l := range keys {
+			if _, err := preader.PostingsOffset(l.Name, l.Value); errors.Is(err, indexheader.NotFoundRangeErr) {
 				// This label name and value doesn't exist in this block, so there are 0 postings we can match.
+				// Set it to an empty value, so we can filter it out later.
+				keys[i] = labels.Label{}
 				// Try with the rest of the set matchers, maybe they can match some series.
 				continue
+			} else if err != nil {
+				retErr = err
+				return nil
+			}
+			existingKeys++
+		}
+		if existingKeys == len(keys) {
+			return keys
+		}
+		if existingKeys == 0 {
+			return nil
+		}
+
+		filtered := make([]labels.Label, 0, existingKeys)
+		for _, k := range keys {
+			if k == (labels.Label{}) {
+				continue
 			}
+			filtered = append(filtered, k)
+		}
+		return filtered
+	}
+
+	return postingGroup{
+		addAll:     g.addAll,
+		addKeys:    filterKeys(g.addKeys),
+		removeKeys: filterKeys(g.removeKeys),
+	}, retErr
+}
+
+type lazyPostingGroup struct {
+	isRemoveGroup bool
+	labelName     string
+	matcher       func(string) bool
+	addAll        bool
+}
+
+func newLazyPostingGroup(addAll bool, isRemoveGroup bool, labelName string, matcher func(string) bool) lazyPostingGroup {
```

Comment:
I don't think we need `isRemoveGroup` as that is always equal to `addAll`.

Reply:
we'd need a different name for it then 🤔 `isNegated`, `isNegativeGroup`, `isExcluded` (as in excludes the results from this)?

Reply:
`isRemoval`? `subtract`? BTW, does it even make sense to have `addKeys` and `removeKeys` when they are mutually exclusive and the fact that we check one or another depends on this flag?

Can we just have something like this?

```go
type postingGroup struct {
    keys []labels.Label
    subtract bool 
}
```

Reply:
i think having separate fields on `postingGroup` makes it easier to work with the group in `expandedPostings()`. But we can unify them within `lazyPostingGroup`.

Reply:
I can't see that it's easier to work with two fields, these are the references in `toPostingGroups`:

```go
		if !pg.addAll && len(pg.addKeys) == 0 {
			return nil, nil, nil
		}
```

:point_up: this one would just be renamed

```go
		hasAdds = hasAdds || len(pg.addKeys) > 0
```

This one would need to add a `!pg.subtract` to the condition, however if we properly filter them, the condition could be just `hasAdds = hasAdds || !pg.subtract` now, since we'd not have `!pg.subtract` with empty keys.

```go
		numKeys += len(pg.addKeys)
		numKeys += len(pg.removeKeys)
```

:point_up: This one would be just one line now, just like this other one:

```go
		keys = append(keys, pg.addKeys...)
		keys = append(keys, pg.removeKeys...)
```

Finally in `expandedPostings` we'd just need to replace `if len(g.addKeys) > 0` condition (which is subtle and indirect, IMO) by an explicit `if !g.subtract`.

Reply:
ok, you're right